RPIPs/RPIP-31.md

```diff
@@ -28,18 +28,45 @@ https://dao.rocketpool.net/t/withdrawal-address-splitter-contract/563
+
+### Defaults
+- The RPL withdrawal address SHALL be initialized as unset for new node operators
+- The RPL withdrawal address SHALL be initialized as unset for existing node operators
+  
+### Rewards & Withdrawn RPL
 - If a node's RPL withdrawal address is set, all RPL rewards and withdrawn RPL will go to the RPL withdrawal address
 - If a node's RPL withdrawal address is unset, all RPL rewards and withdrawn RPL will go to the primary withdrawal address
+
+### Setting the Address
 - As a node operator, I MUST be able to set an RPL withdrawal address so that I can direct my RPL bond/rewards
   - If a node's RPL withdrawal address is unset, the change MUST come from the node's primary withdrawal address
   - If a node's RPL withdrawal address is set, the change MUST come from the current RPL withdrawal address
 - As a node operator, I MUST be able to confirm the RPL withdrawal address from the new target address before it takes effect, to ensure it is the correct address
 - As the owner of the RPL withdrawal address, I MUST be able to unset my RPL withdrawal address
+
+### Withdrawing Excess RPL
 - As the controller of the RPL withdrawn from a node, I MUST be able to call `RocketNodeStaking.withdrawRPL()` to withdraw excess RPL
   - If a node's RPL withdrawal address is unset, the call MUST come from one of: the node's primary withdrawal address, or the node's address
   - If a node's RPL withdrawal address is set, the call MUST come from the current RPL withdrawal address
-- The RPL withdrawal address SHALL be initialized as unset for new node operators
-- The RPL withdrawal address SHALL be initialized as unset for existing node operators
+
+### Claiming RPL rewards
+- As the controller of the RPL for a node, I MUST be able to call `RocketMerkleDistributorMainnet.claim()` to trigger a claim of RPL rewards
+  - If a node's RPL withdrawal address is unset, the call MUST come from one of: the node's primary withdrawal address, or the node's address
+  - If a node's RPL withdrawal address is set, the call MUST come from one of: the node's primary withdrawal address, the current RPL withdrawal address, or the node's address
+- As the controller of the RPL for a node, I MUST be able to call `RocketMerkleDistributorMainnet.claimAndStake()` to trigger a claim of RPL rewards and restake a portion
+  - If a node's RPL withdrawal address is unset, the call MUST come from one of: the node's primary withdrawal address, or the node's address
+  - If a node's RPL withdrawal address is set, the call MUST come from the current RPL withdrawal address
+
+### Staking RPL
+- As the controller of the RPL for a node, I MUST be able to call `RocketNodeStaking.stakeRPL()` to stake additional RPL
+  - If a node's RPL withdrawal address is unset, the call MUST come from the node's address
+  - If a node's RPL withdrawal address is set, the call MUST come from the current RPL withdrawal address
```

Comment:
This is confusing. The node address is no longer allowed to call stakeRPL here? And the RPL withdrawal address can call _either_ stakeRPL _or_ stakeRPLFor with equivalent results?

Reply:
If the RPL withdrawal address is set to a third-party, any RPL staked by the node will be sent to the RPL withdrawal address owner. We would prefer this to be safe-by-default so a node operator cannot fall into this trap. 

I don't really see a case where the node operator would need to stake RPL from the node. If they own the RPL address, it is more convenient to do it from the RPL withdrawal address. If they do not own the RPL address, they are in danger of giving RPL to the withdrawal address owner.

In the draft, I have simply replaced the node operator address with the RPL withdrawal address but we could make that not function inaccessible to the node address, if the RPL withdrawal address is set. That would mean the withdrawal addresses consistently use **stakeRPLFor** instead.

Reply:
I _do_ prefer a single way of doing things to two equivalents (as you mention in the last paragraph).

Insofar as why someone would want to avoid using their RPL withdrawal address, for me that's a cold wallet that's (purposely) a little bit of a pain to use. That said, I think this is addressed by `setStakeRPLForAllowed`

I think that would get us to something like:
```
- As the controller of the RPL for a node, I MUST be able to stake additional RPL
  - If a node's RPL withdrawal address is unset, there MUST be a method to do this from the node's address (eg, by calling `RocketNodeStaking.stakeRPL()`), or a specifically allowed address (eg, by calling `RocketNodeStaking.stakeRPLFor()`)
  - If a node's RPL withdrawal address is set, there MUST be a method to do this from the current RPL withdrawal address (eg, by calling `RocketNodeStaking.stakeRPLFor()`), or a specifically allowed address (eg, by calling `RocketNodeStaking.stakeRPLFor()`)
- As the controller of the RPL for a node, I MUST be able to call to control the list of specifically allowed RPL staking addresses (eg, by calling `RocketNodeStaking.setStakeRPLForAllowed()`)
  - If a node's RPL withdrawal address is unset, the call MUST come from the node's address
  - If a node's RPL withdrawal address is set, the call MUST come from the current RPL withdrawal address
```

Reply:
The only issue with the "there MUST be a method to do this" wording is that it doesn't explicitly say that the node operator cannot stake RPL. I think we can remove the implementation details and it makes the spec clearer. 
How is this?
```
- As the controller of the RPL for a node, I MUST be able to stake additional RPL
  - If a node's RPL withdrawal address is unset, the call MUST come from the node's address, or a specifically allowed address
  - If a node's RPL withdrawal address is set, the call MUST come from the current RPL withdrawal address, or a specifically allowed address
- As the controller of the RPL for a node, I MUST be able to control the list of specifically allowed RPL staking addresses
  - If a node's RPL withdrawal address is unset, the call MUST come from the node's address
  - If a node's RPL withdrawal address is set, the call MUST come from the current RPL withdrawal address
  ```

Reply:
Lgtm